feat(bazel): add dts bundler as action to ng_module

alan-agius4 opened the pull request:
feat(bazel): add dts bundler as action to ng_module

This enabled dts flattening in the final distributable package.

Notes:
 - For the time being this is an opt-in feature via the `ng_module` attribute  `bundle_dts`, however in the near future this will be turned on by default.
 - This only supports the legacy compiler `ngc`, as `ngtsc` emits namespaced imports `import * as __` from local modules which is not supported for the time being by API Extractor. See: https://github.com/Microsoft/web-build-tools/issues/1029

Ref: TOOL-611

alan-agius4 committed:
feat(bazel): add dts bundler as action to ng_module

This enabled dts flattening in the final distrubutable package.

Notes:
 - For the time being this is an opt-in feature via the `ng_module` attribute  `bundle_dts`, however in the near future this will be turned on by default.
 - This only supports the legacy compiler `ngc`, as `ngtsc` emits namespaced imports `import * as __` from local modules which is not supported for the time being by API Extractor. See: https://github.com/Microsoft/web-build-tools/issues/1029

Ref: TOOL-611

alan-agius4 committed:
test: update ng_package tests to use bundle_dts

alexeagle commented on packages/bazel/src/api-extractor/BUILD.bazel:
```diff
@@ -0,0 +1,28 @@
+package(default_visibility = ["//visibility:public"])
```
we don't want to make all rules visible everywhere. This is the wrong choice already made in many angular packages, but we shouldn't make it worse.

add visibility to the `api_extractor` target only (no one depends on the ts_library right?) and make it specific to what package should use it

`visibility = ["//packages:__subpackages__"]` I think

alexeagle commented on packages/bazel/src/api-extractor/index.ts:
```diff
@@ -0,0 +1,112 @@
+
+export function runMain(
+    tsConfig: string, entryPoint: string, dtsBundleOut?: string, apiReviewFolder?: string,
+    acceptApiUpdates = false): number {
```
nit: instead of `number` the better return type is probably `0|1` ?

alan-agius4 replied:
nice catch

alexeagle commented on packages/bazel/src/api-extractor/tsconfig.json:
```diff
@@ -0,0 +1,12 @@
+{
```
can you avoid creating a new `tsconfig.json`? we have no good way to keep settings in sync.

If you need typings or libs in the program, use `/// <reference types="node"/>` or `/// <reference lib="es2017"/>` in the TS file

alexeagle commented on packages/bazel/src/ng_module.bzl:
```diff
@@ -85,7 +86,7 @@ def _is_legacy_ngc(ctx):
 
     Returns:
-      true iff the current compilation strategy will produce View Engine compilation outputs (such as
+      true if the current compilation strategy will produce View Engine compilation outputs (such as
```
nit: iff was intentional I think, as a shorthand for "if and only if"

alexeagle commented on packages/bazel/src/ng_module.bzl:
```diff
@@ -543,6 +598,11 @@ NG_MODULE_ATTRIBUTES = {
         executable = True,
         cfg = "host",
     ),
+    "api_extractor": attr.label(
```
prefix it with underscore so that users can't set this attribute - not meant to be public API

alexeagle commented on packages/bazel/src/ng_package/ng_package.bzl:
```diff
@@ -235,6 +229,22 @@ def _ng_package_impl(ctx):
+
+        if len(type_definitions) > 0 and len(bundled_type_definitions) > 0:
+            # bundle_dts needs to be enabled/disabled for all ng module packages.
+            fail("Expected all or none deps targets to have 'bundle_dts' enabled.")
```
can you reword this so it makes more sense?

alexeagle commented on packages/bazel/src/ng_package/packager.ts:
```diff
@@ -176,7 +178,30 @@ function main(args: string[]): number {
+
+  dtsBundles.forEach(bundleFile => {
+    const cleanDistPath = bundleFile.replace(dtsBundleFileSuffix, '.d.ts');
+    // API extractor will not dedupe license comments from various files
```
this seems risky, what if there is a third-party license included? for example tslib?
I think we should only strip duplicates, not assume that there is only a single license to apply at the top, unless we have a strong argument for why we can assume that all the code is under our license

alan-agius4 replied:
I assume that all the code is ours because the api extractor will bundle local dts files. Ie with relative imports.

alexeagle commented on packages/bazel/src/ng_package/packager.ts:
```diff
@@ -368,6 +391,39 @@ export * from '${srcDirRelative(inputPath, typingsFile.replace(/\.d\.tsx?$/, '')
+  * @param metadataPath the metadata file path
+  * @param typingsPath the typings bundle entrypoint
+  */
+  function rewireMetadata(metadataPath: string, typingsPath: string): string {
```
I'd like to see a unit test for this function. Maybe need to add a new spec if we don't have one already

alan-agius4 replied:
I didn't add a unit test, because there is the golden file test 
https://github.com/angular/angular/pull/28588/files#diff-97fd23e5d431eaf958713d06651edcb7R496

But if you prefer i can add some tests for this.

alexeagle reviewed (commented):
Awesome, this looks really close

alan-agius4 committed:
fixup! feat(bazel): add dts bundler as action to ng_module

alan-agius4 committed:
build: update `@microsoft/api-extractor` to `7.0.17`

This fixes one of the problems that we reported https://github.com/Microsoft/web-build-tools/issues/1048

alexeagle commented:
[presubmit](http://test/OCL:233615649:BASE:233615815:1549990826565:a5cb20ae)

alexeagle committed:
build: make api-extractor work in google3

alexeagle commented:
fixed? [presubmit](http://test/OCL:233615649:BASE:233685364:1550035159870:1731a142)

alexeagle commented:
still failing in g3: `packages/bazel/src/ng_module.bzl", line 367, in ngc_compile_action
		[("--node_options=%s" % opt) for opt in node_opts]
type 'string' is not iterable. You may still use `len` and string indexing. Use --incompatible_string_is_not_iterable=false to temporarily disable this check.`

alan-agius4 commented:
I might know why this is failing in G3.

I’ll try to land a fix in an hr.

On Wed, 13 Feb 2019 at 16:26, Alex Eagle <notifications@github.com> wrote:

> still failing in g3: packages/bazel/src/ng_module.bzl", line 367, in
> ngc_compile_action [("--node_options=%s" % opt) for opt in node_opts] type
> 'string' is not iterable. You may still use len and string indexing. Use
> --incompatible_string_is_not_iterable=false to temporarily disable this
> check.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/28588#issuecomment-463240802>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQv-WnewqVu2vDiIFCFBs8s6i4s8Z4mjks5vNC63gaJpZM4am9z->
> .
>

alan-agius4 committed:
fixup! build: make api-extractor work in google3

alexeagle commented:
Caretaker: the last [presubmit](https://test.corp.google.com/ui#id=OCL:233615649:BASE:233787608:1550084303629:41833148) was GREEN
please merge
(I'm not on corp to set the status myself)
